### DISCUSSION TITLE
SharedArrayBuffer

### DISCUSSION THREAD

**Discussion by tcf909** [uNetworking/uWebSockets.js#493] [Ideas]
I'm doing some early work around the use of uWebSockets in worker_threads. I'm doing a multicast like setup that will allow multiple workers to listen for messages (among other things).

Any thoughts on proving SharedArrayBuffers rather than ArrayBuffers if we option for it?

I think this could be a much more common need as developers evolve and start using more worker_thread oriented code.

Right now I'm converting to a SharedArrayBuffer if there is more than one interested party using a new SharedArrayBuffer instance along with a TypeArray instance (to copy the data with .set() ).

A single copy is made which does cut down the overhead (vs many copies), but it could be zero copies (lower latency) if the initial data was already in a SharedArrayBuffer.

**Comment by ghost** [comment]
You have to copy data once anyways, the ArrayBuffer is just a pointer to internal memory. That memory is not valid once you return from the callback, and certainly not for other threads. So copying to a SharedArrayBuffer if you need it, is already minimal.

**Reply by tcf909** [reply]
If I transfer the ArrayBuffer provided by uWebSocket it seems to work ok?

```
messagePort.postMessage(arrayBuffer, [arrayBuffer]);
```

Which is faster from what I can tell than copying to a shared array buffer and transferring the pointer of the shared array buffer.

The data seems available in the other thread when I just transfer arrayBuffer. I'm wondering what I'm missing here?

Does the ArrayBuffer somehow still get wiped even though it has been transferred? (I'm not observing this with basic testing if so)

**Reply by ghost** [reply]
Like I said, the ArrayBuffer is just a pointer to internal memory. That memory is not valid once you return from the callback
